File: examples/basic_usage.cpp
#include <iostream>
#include <filesystem>

#include "sys-sage.hpp"

using namespace sys_sage;

using std::cout;
using std::endl;

void usage(char* argv0)
{
    std::cerr << "usage: " << argv0 << " <hwloc xml path> <caps-numa-benchmark csv path>" << std::endl;
    std::cerr << "       or" << std::endl;
    std::cerr << "       " << argv0 << " (uses predefined paths which may be incorrect.)" << std::endl;
    return;
}

int main(int argc, char *argv[])
{
    std::string xmlPath;
    std::string bwPath;
    if(argc < 2){
        xmlPath = EXAMPLE_DIR + std::string("/skylake_hwloc.xml");
        bwPath = EXAMPLE_DIR + std::string("/skylake_caps_numa_benchmark.csv");
    }
    else if(argc == 3){
        xmlPath = argv[1];
        bwPath = argv[2];
    }
    else{
        usage(argv[0]);
        return 1;
    }

    //create root Topology and one node
    Topology* topo = new Topology();
    Node* n = new Node(topo, 1);

    
    cout << "-- Parsing Hwloc output from file " << xmlPath << endl;
    if(parseHwlocOutput(n, xmlPath) != 0) { //adds topo to a next node
        usage(argv[0]);
        return 1;
    }
    cout << "-- End parseHwlocOutput" << endl;

    cout << "Total num HW threads: " << topo->CountDescendantsByType(sys_sage::ComponentType::Thread) << endl;

    cout << "---------------- Printing the whole tree ----------------" << endl;
    topo->PrintSubtree();
    cout << "----------------                     ----------------" << endl;
    cout << "-- Parsing CapsNumaBenchmark output from file " << bwPath << endl;
    if(parseCapsNumaBenchmark((Component*)n, bwPath, ";") != 0){
        cout << "failed parsing caps-numa-benchmark" << endl;
        usage(argv[0]);
        return 1;
    }
    cout << "-- End parseCapsNumaBenchmark" << endl;

    cout << "---------------- Printing all Relations ----------------" << endl;
    n->PrintRelationsInSubtree();
    cout << "----------------                        ----------------" << endl;

    std::string output_name = "sys-sage_sample_output.xml";
    cout << "-------- Exporting as XML to " << output_name << " --------" << endl;
    exportToXml(topo, output_name);

    cout << "--------Clearing up the tree--------" << endl;
    Component::DeleteSubtree(topo);
    return 0;
}

File: examples/basic_usage.py
import sys_sage as syge 
import sys
import os

def usage(argv0):
    print("usage: {} <hwloc xml path> <caps-numa-benchmark csv path>".format(argv0), file=sys.stderr)
    print("       or", file=sys.stderr)
    print("       {} (uses predefined paths which may be incorrect.)".format(argv0), file=sys.stderr)
    return

def main():
    if len(sys.argv) < 2:
        path_prefix = os.path.dirname(os.path.abspath(sys.argv[0]))
        xmlPath = os.path.join(path_prefix, "example_data", "skylake_hwloc.xml")
        bwPath = os.path.join(path_prefix, "example_data", "skylake_caps_numa_benchmark.csv")
    elif len(sys.argv) == 3:
        xmlPath = sys.argv[1]
        bwPath = sys.argv[2]
    else:
        usage(sys.argv[0])
        return 1

    #create root Topology and one node
    topo = syge.Topology()
    n = syge.Node(topo, 1)

    print("-- Parsing Hwloc output from file", xmlPath)
    if syge.parseHwlocOutput(n, xmlPath) != 0: #adds topo to a next node
        usage(sys.argv[0])
        return 1
    print("-- End parseHwlocOutput")

    print("Total num HW threads:", topo.CountAllSubcomponentsByType(syge.COMPONENT_THREAD))
    print("---------------- Printing the whole tree ----------------")
    topo.PrintSubtree()
    print("----------------                     ----------------")

    print("-- Parsing CapsNumaBenchmark output from file", bwPath)
    if syge.parseCapsNumaBenchmark(n, bwPath, ";") != 0:
        print("failed parsing caps-numa-benchmark")
        usage(sys.argv[0])
        return 1
    print("-- End parseCapsNumaBenchmark")

    print("---------------- Printing all DataPaths ----------------")
    n.PrintAllDataPathsInSubtree()
    print("----------------                        ----------------")

    output_name = "sys-sage_sample_output.xml"
    print("-------- Exporting as XML to", output_name, "--------")
    syge.exportToXml(topo, output_name)

    print("--------Clearing up the tree--------")
    topo.Delete(True)

    return 0

if __name__ == "__main__":
    main()